src/nvim/buffer_defs.h

```diff
@@ -1292,8 +1292,9 @@ struct window_S {
   linenr_T w_stl_line_count;         // line count when last redrawn
   int w_stl_topfill;                 // topfill when last redrawn
   char w_stl_empty;                  // true if elements show 0-1 (empty line)
-  int w_stl_state;                   // State when last redrawn
   int w_stl_recording;               // reg_recording when last redrawn
+  int w_stl_state;                   // get_real_state() when last redrawn
+  int w_stl_visual_mode;             // VIsual_mode when last redrawn
```

Comment:
Might be able to get away with just adding `VIsual_mode` to `state` instead of adding another field:
```diff
@@ -823,7 +823,7 @@ void setcursor_mayforce(bool force)
 /// @param always  if false, only show ruler if position has changed.
 void show_cursor_info_later(bool force)
 {
-  int state = get_real_state();
+  int state = get_real_state() + VIsual_mode;
   int empty_line = (State & MODE_INSERT) == 0
                    && *ml_get_buf(curwin->w_buffer, curwin->w_cursor.lnum, false) == NUL;

@@ -839,8 +839,7 @@ void show_cursor_info_later(bool force)
       || curwin->w_topfill != curwin->w_stl_topfill
       || empty_line != curwin->w_stl_empty
       || reg_recording != curwin->w_stl_recording
-      || state != curwin->w_stl_state
-      || (VIsual_active && VIsual_mode != curwin->w_stl_visual_mode)) {
+      || state != curwin->w_stl_state) {
     if (curwin->w_status_height || global_stl_height()) {
       curwin->w_redr_status = true;
     } else {
@@ -865,9 +864,6 @@ void show_cursor_info_later(bool force)
   curwin->w_stl_topfill = curwin->w_topfill;
   curwin->w_stl_recording = reg_recording;
   curwin->w_stl_state = state;
-  if (VIsual_active) {
-    curwin->w_stl_visual_mode = VIsual_mode;
-  }
 }

 /// @return true when postponing displaying the mode message: when not redrawing
```

Reply:
I don't think this makes sense. I think it's better to combine `VIsual_mode` with other small fields like `reg_recording`, but what's wrong with adding another field?

Reply:
Nothing per se, just thought since both fields combined determine the actual state, we might as well combine them.